lib/AstOfLlbc.ml

```diff
@@ -1025,57 +1026,68 @@ let rec expression_of_fn_ptr env depth (fn_ptr: C.fn_ptr) =
-
-            (* This must be in agreement, and in the same order as build_trait_clause_mapping *)
-            List.map (fun (item_name, decl_id) ->
-              let fn_ptr: C.fn_ptr = {
-                func = TraitMethod (trait_ref, item_name, decl_id);
-                generics = Charon.TypesUtils.empty_generic_args
-              } in
-              let fn_ptr = fst3 (expression_of_fn_ptr env (depth ^ "  ") fn_ptr) in
-              fn_ptr
-            ) (trait_impl.required_methods @ trait_impl.provided_methods)
-
-        | Clause clause_id ->
-            (* Caller it itself polymorphic and refers to one of its own clauses to synthesize
-               the clause arguments at call-site. *)
-            List.rev (Krml.KList.filter_mapi (fun i (var, t) ->
-              match var with
-              | TraitClauseMethod (clause_id', _, _) when clause_id = clause_id' ->
-                  Some K.(with_type t (EBound i))
-              | _ ->
-                  None
-            ) env.binders)
-
-        | ParentClause (_instance_id, decl_id, clause_id) ->
-            let trait_decl = env.get_nth_trait_decl decl_id in
-            let name = string_of_name env trait_decl.name in
-            let clause_id = C.TraitClauseId.to_int clause_id in
-            let parent_clause = List.nth trait_decl.parent_clauses clause_id in
-            let parent_clause_decl = env.get_nth_trait_decl parent_clause.trait_id in
-            let parent_name = string_of_name env parent_clause_decl.name in
-            Krml.KPrint.bprintf "looking up parent clause #%d of decl=%s = %s\n" clause_id name
-              parent_name;
-            if List.mem parent_name blocklisted_trait_decls then
+          match trait_ref.trait_id with
+          | _ when List.mem name blocklisted_trait_decls ->
+              (* Trait not supported -- don't synthesize arguments *)
               []
-            else
-              failwith ("Don't know how to resolve trait_ref " ^ C.show_trait_ref trait_ref)
 
-        | _ ->
-            failwith ("Don't know how to resolve trait_ref " ^ C.show_trait_ref trait_ref)
+          | TraitImpl impl_id ->
+              (* Call-site has resolved trait clauses into a concrete trait implementation. *)
+              let trait_impl: C.trait_impl = env.get_nth_trait_impl impl_id in
+
+              (* This must be in agreement, and in the same order as build_trait_clause_mapping *)
+              List.map (fun (item_name, decl_id) ->
+                let fn_ptr: C.fn_ptr = {
+                  func = TraitMethod (trait_ref, item_name, decl_id);
+                  generics = Charon.TypesUtils.empty_generic_args
+                } in
+                let fn_ptr = fst3 (expression_of_fn_ptr env (depth ^ "  ") fn_ptr) in
+                fn_ptr
+              ) (trait_impl.required_methods @ trait_impl.provided_methods) @
+              build_trait_ref_mapping ("  " ^ depth) trait_ref.generics.trait_refs
```

Comment:
here, the idea is that after filling out required and provided methods, this should match what is done in build_trait_clause_mapping for the parent clauses -- but I'm not sure this is the right way, so please check!

Reply:
So the assumption is that the list of trait_refs clauses matches the required list of clause arguments to the trait? That feels correct. I've been wanting to provide a better API for things like that.

Reply:
ok, well I believe this should provide a good use-case for that :-)